include/envoy/thread_local/thread_local.h

```diff
@@ -68,17 +48,53 @@ class Instance {
 
   /**
    * Set thread local data on all threads previously registered via registerThread().
-   * @param index species the slot ID which is used in subsequent calls to get() and getTyped().
    * @param initializeCb supplies the functor that will be called *on each thread*. The functor
    *                     returns the thread local object which is then stored. The storage is via
    *                     a shared_ptr. Thus, this is a flexible mechanism that can be used to share
    *                     the same data across all threads or to share different data on each thread.
    */
   typedef std::function<ThreadLocalObjectSharedPtr(Event::Dispatcher& dispatcher)> InitializeCb;
-  virtual void set(uint32_t index, InitializeCb cb) PURE;
+  virtual void set(InitializeCb cb) PURE;
+};
+
+typedef std::unique_ptr<Slot> SlotPtr;
+
+/**
+ * Interface used to allocate thread local slots.
+ */
+class SlotAllocator {
+public:
+  virtual ~SlotAllocator() {}
+
+  /**
+   * @return SlotPtr a dedicated slot for use in further calls to get(), set(), etc.
+   */
+  virtual SlotPtr allocateSlot() PURE;
+};
+
+/**
+ * Interface for getting and setting thread local data as well as registering a thread
+ */
+class Instance : public SlotAllocator {
+public:
+  /**
+   * A thread (via its dispatcher) must be registered before set() is called on any allocated slots
+   * to receive thread local data updates.
+   * @param dispatcher supplies the thread's dispatcher.
+   * @param main_thread supplies whether this is the main program thread or a worker thread.
```

Comment:
Right now, no. Though, in reality nothing in TLS requires it to be a "worker." It can just be anything with a dispatcher that can receive a post. The only things that register right now are workers though.